src/NHibernate/Engine/Query/HQLQueryPlan.cs

```diff
@@ -174,6 +198,69 @@ public IEnumerable PerformIterate(QueryParameters queryParameters, IEventSource
 			return new JoinedEnumerable(results);
 		}
 
+		// Since v5.3
+		[Obsolete("This method has no more usages and will be removed in a future version")]
+		public async Task<IEnumerable> PerformIterateAsync(QueryParameters queryParameters, IEventSource session, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			if (Log.IsDebugEnabled())
+			{
+				Log.Debug("enumerable: {0}", _sourceQuery);
+				queryParameters.LogParameters(session.Factory);
+			}
+			if (Translators.Length == 0)
+			{
+				return CollectionHelper.EmptyEnumerable;
+			}
+			if (Translators.Length == 1)
+			{
+				return await (Translators[0].GetEnumerableAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false);
+			}
+			var results = new IEnumerable[Translators.Length];
+			for (int i = 0; i < Translators.Length; i++)
+			{
+				var result = await (Translators[i].GetEnumerableAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false);
+				results[i] = result;
+			}
+			return new JoinedEnumerable(results);
+		}
+
+		// Since v5.3
+		[Obsolete("This method has no more usages and will be removed in a future version")]
+		public async Task<IEnumerable<T>> PerformIterateAsync<T>(QueryParameters queryParameters, IEventSource session, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			return new SafetyEnumerable<T>(await (PerformIterateAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false));
+		}
+
+		public IAsyncEnumerable<T> PerformAsyncIterate<T>(QueryParameters queryParameters, IEventSource session)
+		{
+			if (Log.IsDebugEnabled())
+			{
+				Log.Debug("enumerable: {0}", _sourceQuery);
+				queryParameters.LogParameters(session.Factory);
+			}
+
+			if (Translators.Length == 0)
+			{
+				return new CollectionHelper.EmptyAsyncEnumerable<T>();
+			}
+
+			if (Translators.Length == 1)
+			{
+				return Translators[0].GetAsyncEnumerable<T>(queryParameters, session);
+			}
+
+			var results = new IAsyncEnumerable<T>[Translators.Length];
+			for (int i = 0; i < Translators.Length; i++)
```

Comment:
Non-async version in case multiple translators opens all DB readers before processing. In case multiple open readers are not supported (like MS SQL Server with MARS disabled by default) - all data populated at once. It makes `PerformIterate` quite useless for such queries (see #2017).

Just want to clarify async behavior. So how async version behaves in this regard? Does it also opens all readers at once?

Reply:
Currently as it is implemented it opens the connection and starts reading only when you call MoveNextAsync first time.